raiden/network/transport/matrix/transport.py

```diff
@@ -1178,20 +1178,38 @@ def _get_user(self, user: Union[User, str]) -> User:
         All users are supposed to be in discovery room, so just reuse the
         `_members` dict already present from the Matrix SDK.
         """
-        user_id: str = getattr(user, "user_id", user)
+        if isinstance(user_or_userid, User):
+            user = user_or_userid
+            user_id = user.user_id
+            displayname = user.displayname
+        else:
+            user = None
+            user_id = user_or_userid
+            displayname = None
+
+        assert user_id, "user_id must be available"
+
         discovery_room = self._broadcast_rooms.get(
             make_room_alias(self.chain_id, DISCOVERY_DEFAULT_ROOM)
         )
-        if discovery_room and user_id in discovery_room._members:
-            duser = discovery_room._members[user_id]
-            # if handed a User instance with displayname set, update the discovery room cache
-            if getattr(user, "displayname", None):
-                assert isinstance(user, User)
-                duser.displayname = user.displayname
-            user = duser
-        elif not isinstance(user, User):
-            user = self._client.get_user(user_id)
-        return user
+
+        # There is no discovery room to reuse the cache, so just return the
+        # user. This happens when the cache is used before the transport is
+        # fully initialized, e.g. by the `UserAddressManager` thread.
+        if discovery_room is None:
+            return user or self._client.get_user(user_id)
+
+        cached_user = discovery_room._members.setdefault(user_id, user)
```

Comment:
I'd agree with @ulope here -- this adds a secondary pathway into the discovery room, that does not correspond to what's happening server side.

Reply:
```suggestion
        if user_id not in discovery_room._members:
            log.warning("Querying _get_user() for user that is unknown online.", user_id=user_id)
        cached_user = discovery_room._members.setdefault(user_id, user)
```

Reply:
> I don't think using setdefault is a good idea here since it will make it look as if a user is a member of the discovery room that really isn't.

Then I'm going to move the cache somewhere else. It doesn't make sense to have a cache that works sometimes but not other times.

Reply:
@ulope can this be a source of bugs? If it can, then I'm totally going to revert the change.

Reply:
> It doesn't make sense to have a cache that works sometimes but not other times.

That case should never happen (except maybe really early in the transport lifecycle), all nodes must be members of the discovery room.

Reply:
> **the** discovery room.

there must be only one.

Reply:
> can this be a source of bugs? If it can, then I'm totally going to revert the change.

You mean modifying the members list? I'm not sure, would have to look more closely which places this is used. But modifying an internal datastructure of another object just seems wrong.

Reply:
I changed the code to use an explicit cache instead of relying in the private member of another object.